File: src/main/java/org/example/kalkulationsprogramm/domain/Kategorie.java
package org.example.kalkulationsprogramm.domain;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;

import java.util.HashSet;
import java.util.Set;

@Getter
@Setter
@Entity
@Table(name = "kategorie")
@Inheritance(strategy = InheritanceType.JOINED)
public class Kategorie
{
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    private String beschreibung;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_kategorie_id")
    private Kategorie parentKategorie;

    /**
     * Typische Liefer-Rollen dieser Kategorie. Leer = erbt von der Oberkategorie.
     * Steuert den Lieferanten-Vorschlag beim Preis-Eintragen am Artikel.
     */
    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "kategorie_rollen", joinColumns = @JoinColumn(name = "kategorie_id"))
    @Column(name = "rolle")
    @Enumerated(EnumType.STRING)
    private Set<LieferantRolle> typischeRollen = new HashSet<>();
}

File: src/main/kotlin/org/example/kalkulationsprogramm/domain/Kategorie.kt
package org.example.kalkulationsprogramm.domain

import jakarta.persistence.*

@Entity
@Table(name = "kategorie")
@Inheritance(strategy = InheritanceType.JOINED)
open class Kategorie {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    open var id: Int? = null

    open var beschreibung: String? = null

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_kategorie_id")
    open var parentKategorie: Kategorie? = null

    @ElementCollection(fetch = FetchType.EAGER)
    @CollectionTable(name = "kategorie_rollen", joinColumns = [JoinColumn(name = "kategorie_id")])
    @Column(name = "rolle")
    @Enumerated(EnumType.STRING)
    open var typischeRollen: MutableSet<LieferantRolle> = mutableSetOf()
}
